### ISSUE TITLE
How to prevent NVDA to overwrite touchscreen configurations on Windows

### ISSUE THREAD

**Issue by simonesilva** [nvaccess/nvda#9682]
### Steps to reproduce:
When NVDA is on, it overwrites a set of predefined touch and gestures configurations for touchscreens, so what was before a single tap action becomes a fast double tap one. I need to prevent NVDA from overwriting some default settings on Windows.

I already tried removing "tap (Touch screen)" from "Input Gestures" under "Preferences" option, but it only stopped dictating, I still have to double tap to perform an action.

### System configuration
I'm using windows 10

#### NVDA version:
2019.1.1

#### Windows version:
Windows 10

**Comment by DrSooom** [comment]
What happens if you press NVDA+F2 before performing the touch gesture?

**Comment by alejoceballos-tw** [comment]
Hi @DrSooom! I believe the issue here is that there ain't a keyboard to press F2. What @simonesilva is asking is to disable double tap feature and keep it disabled when restarting Windows in a touch only device.

By-the-way, what is the "NVDA" from "NVDA+F2", is it a special key from a specific device?

**Comment by XLTechie** [comment]
On Mon, 10 Jun 2019, Alejo Ceballos wrote:

> By-the-way, what is the "NVDA" from "NVDA+F2", is it a special key from a specific device?

A screenreader needs to handle many commands from the user--to read lines, 
characters, words, windows, etc.; to move to different parts of the screen; to 
change how certain information is being presented because more or less detail is 
needed; and dozens of more things. So a special modifier key must be used, since 
the alt, control, and other such keys are already used by most applications.

Since we're talking about the NVDA screen reader here, that key is called 
"NVDA". It is usually the insert key on the keyboard, and most users also remap 
the capslock key to serve as an additional NVDA key.

So, for example, pressing NVDA+1, which just means capslock and the number 1, 
enters key describer mode, wherein NVDA describes what all the other keys and 
key combinations do. NVDA+q exits the screenreader. NVDA+shift+g is a quick 
shortcut to open the screenreader's general settings dialog. NVDA+n brings up 
the screenreader main menu. And so on.

**Comment by XLTechie** [comment]
@simonesilva Can you explain more of the use case here?

1. Is this a touch-only device, as @alejoceballos-tw said?
    * If so: How does the user expect to explore the screen by touch, if every touch activates something? That is the purpose of double tap in this context.
2. If it is not a touch-only device, is it your intention to have the screen be used by sighted users--I.E. someone not depending on NVDA, while the NVDA user works only by using the keyboard?
3. If not 1 or 2, then what is your use case for NVDA here?

**Comment by josephsl** [comment]
Hi, for use case 1, already resolved (at least partially): Enhanced Touch Gestures add-on does have touch passthrough mode where one can ask NVDA to deactivate its own touchscreen support (yours truly wrote that back in 2017; I know that I start to sound like a salesperson). Thanks.

**Comment by XLTechie** [comment]
@simonesilva Can you explain more of the use case here?

1. Is this a touch-only device, as @alejoceballos-tw said?
    * If so: How does the user expect to explore the screen by touch, if every touch activates something? That is the purpose of double tap in this context.
2. If it is not a touch-only device, is it your intention to have the screen be used by sighted users--I.E. someone not depending on NVDA, while the NVDA user works only by using the keyboard?
3. If not 1 or 2, then what is your use case for NVDA here?

**Comment by XLTechie** [comment]
@josephsl wrote:
> Hi, for use case 1, already resolved (at least partially): Enhanced Touch Gestures add-on does have touch passthrough mode where one can ask NVDA to deactivate its own touchscreen support (yours truly wrote that back in 2017; I know that I start to sound like a salesperson).

Hey if you wrote it, sell it baby!
That might be the groundwork for @simonesilva 's solution, although I still wonder about the use-case for disabling it globally on boot, if in fact that's what @simonesilva wanted.

Anyway, here, I'll help you sell: https://addons.nvda-project.org/addons/enhancedTouchGestures.en.html

**Comment by DrSooom** [comment]
Starting NVDA with the command line parameter/flag "--no-sr-flag" isn't a good solution here. I'm able to see some situations, where multiple users are working together on one single machine. Therefore it makes sense to add a new option in NVDA to completely disable the NVDA's touch functionality via a simple checkbox in the NVDA Touch Interaction Settings as well as via a new flag called "--disable-touch".

@josephsl: Sorry for the stupid question, but are there some specific reasons, why your add-on isn't included into NVDA yet? Or is it already on the way? (I'm not familiar with the whole touch feature in NVDA and on Windows 10. Well, and I don't have a touch screen here too.)

**Comment by josephsl** [comment]
Hi,

Including Enhanced Touch Gestures to Core: that's a long-term plan.

One can disable touch support by running a portable copy of NVDA, but that's not ideal for some scenarios. You can disable touchscreen support after NVDA loads, and that's the mechanism behind the solution in my add-on.

I'll add a way to do this globally in the next version of Enhanced Touch Gestures.

Thanks.

**Comment by Qchristensen** [comment]
From a conversation with a user, ideally as and when such functionality is available in NVDA itself, it would be useful to be able to set it in configuration profiles, enabling it to be set globally, or per app.

**Comment by josephsl** [comment]
Hi,

It'll take me a day or two to come up with a pull request that will resolve this once and for all, and before writing that, I too am interested in hearing back from the original poster as to why this feature is needed (I think I know why, but just want to make sure this person clarifies things).

There is a more important touchscreen PR, and that is right-mouse click simulation. Both PR's are independent from one another, but I'd give slightly higher priority to mouse click simulation.

Thanks.

**Comment by donalfitzpatrick** [comment]
Hi,

Would like to add a use-case for this if I may.  We've developed a web app which enables editing and/or exploration of mathematics.  Part of our functionality is to enable a user to explore the shape of graphs on a touchscreen device.  The ability to have NVDA ignore all gestures on a per-app, or indeed in specific screens in a given app, would be very useful.

**Comment by josephsl** [comment]
Hi, although disabling touch support for specific parts of an app may not come to fruition for a while, at least disabling from apps as a whole will be possible. Thanks.

**Comment by josephsl** [comment]
Hi,

Sometime this week I'll introduce a prototype try build that will demonstrate what the fix will look like (note that the try build must be installed in order for the fix to work).

Thanks.

**Comment by josephsl** [comment]
Hi,

Please try this [try build](https://ci.appveyor.com/api/buildjobs/4w67saj11tpcpnlt/artifacts/output%2Fnvda_snapshot_try-touchSupportToggle-18867%2C7b400064.exe) and see if it helps. Note that you must be running this installed on a touchscreen computer, and it does not support profile switching yet. I'll introduce profile switching support and a command to enable/disable touchscreen support (unassigned) once this try build is verified as working.

Thanks.

**Comment by Deon588** [comment]
Thought i'd add my use case for this feature:  I am a low vision user and use NVDA for reading text however there are lots of other interactions e.g. closing/dragging a window that I would like to be able to do using the normal windows touch interface as I have enough site to perform thos actions without the NVDA overlay.  Would be amazing if this could be toggled by a keyboard shortcut....

**Comment by josephsl** [comment]
Hi, already possible (sort of) via an add-on called Enhanced Touch Gestures. The try build I posted yesterday brings some of that add-on’s functionality to NVDA itself, which is what is requested in this issue. Thanks.

**Comment by leonardder** [comment]
@josephsl: Are you intending to provide a pr for this shortly?

**Comment by josephsl** [comment]
Hi, yes, but not until after 2019.3 stable version is released. Thanks.

**Comment by comanna** [comment]
Hi,

 

Can Cortana and NVDA work together to allow voice control of Windows 10 on the PC or laptop?

 

 

All the best,

 

Cearbhall

 

m +353 (0)833323487 Ph: _353 (0)1-2864623 e: cearbhall.omeadhra@blbc.ie

**Comment by josephsl** [comment]
Hi, in a way, already possible (let’s discuss this on a separate issue). Thanks.

**Comment by josephsl** [comment]
Hi,

And now, with 2019.3 under translatable string freeze and with first 2020.1 changelog entry showing, it is time for a pull request (coming up shortly).

I'm aware that some folks might be asking for a shortcut command to toggle touch support. For now that won't be part of the pull request - to be evaluated at a later time. An ideal solution for this issue would be a touch-based passthrough command.

Thanks.
